clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7543,10 +7541,20 @@ static void handleSYCLIntelMemoryAttr(Sema &S, Decl *D,
     }
   }
 
-  // We are adding a user memory attribute, drop any implicit default.
-  if (auto *MA = D->getAttr<SYCLIntelMemoryAttr>())
-    if (MA->isImplicit())
-      D->dropAttr<SYCLIntelMemoryAttr>();
+  if (auto *MA = D->getAttr<SYCLIntelMemoryAttr>()) {
+    // Check to see if there's a duplicate memory attribute with different
+    // values already applied to the declaration.
+    if (!MA->isImplicit()) {
+      if (MA && MA->getKind() != Kind) {
```

Comment:
You don't need to check MA for NULL here, the if condition above guarantees it.
```suggestion
      if (MA->getKind() != Kind) {
```

Reply:
If getNumArgs() != 0, you can get here with `Kind` not being set.   I am not sure if the check above is valid in that case.

Reply:
> If getNumArgs() != 0, you can get here with `Kind` not being set. I am not sure if the check above is valid in that case.

@premanandrao, did you mean to say getNumArgs() == 0?
```
if (AL.getNumArgs() == 0)
    Kind = SYCLIntelMemoryAttr::Default;
```
The patch generates no diagnostic message when kind is not being set.

```
// Check to see if there's a duplicate attribute with same Default values
  // already applied to the declaration. Drop the duplicate attribute. No diagnostic is generated. 
  // CHECK: VarDecl{{.*}}mem_memory
  // CHECK: SYCLIntelMemoryAttr{{.*}}Default
  // CHECK-NOT: SYCLIntelMemoryAttr
  [[intel::fpga_memory]]
  [[intel::fpga_memory]] unsigned int mem_memory[64];

```

Reply:
>>If getNumArgs() != 0, you can get here with Kind not being set. I am not sure if the check above is valid in that case.

I have updated patch to cover this case with commit https://github.com/intel/llvm/pull/12175/commits/cce91574d1a34ee2d3a2d4a8ea484c3fc1e2a66c

@premanandrao does this answer your concerns?

Reply:
I am not sure.

  ```
if (AL.getNumArgs() == 0)
    Kind = SYCLIntelMemoryAttr::Default;
  else {
    StringRef Str;
    if (!S.checkStringLiteralArgumentAttr(AL, 0, Str))
      return;
    if (Str.empty() ||
        !SYCLIntelMemoryAttr::ConvertStrToMemoryKind(Str, Kind)) {
      SmallString<256> ValidStrings;
      SYCLIntelMemoryAttr::generateValidStrings(ValidStrings);
      S.Diag(AL.getLoc(), diag::err_intel_fpga_memory_arg_invalid)
          << AL << ValidStrings;
      return;
    }
  }
```
When getNumArgs is 0, `Kind` is set to `Default`.   In the else clause, Kind is not set.   So when you compare `getKind()` with `Kind`, what are you checking?

Reply:
Doing more investigation, it turns out the patch below would cover all cases. We do not need to check explicitly getNumArgs() and Kind value.

1. When getNumArgs() is 0, Kind is being set to SYCLIntelMemoryAttr::Default
2. In the else clause, Kind is not set (When getNumArgs() != 0), Kind is being set here to SYCLIntelMemoryAttr::MLAB || SYCLIntelMemoryAttr::BlockRAM

```
if (!MA->isImplicit()) {
+      if (MA->getKind() != Kind) {
+       S.Diag(AL.getLoc(), diag::warn_duplicate_attribute) << &AL;
+        S.Diag(MA->getLocation(), diag::note_previous_attribute);
+      }
+      // Drop the duplicate attribute.
+      return;
+    }
+    // We are adding a user memory attribute, drop any implicit default.
+    D->dropAttr<SYCLIntelMemoryAttr>();
+  }

```

Reply:
Thanks for checking.